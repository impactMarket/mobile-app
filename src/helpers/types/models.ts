// same as in api/types/models

import { UbiCommunityContract } from './ubi/ubiCommunityContract';
import { UbiCommunityDailyMetrics } from './ubi/ubiCommunityDailyMetrics';
import { UbiCommunityState } from './ubi/ubiCommunityState';
import { UbiOrganization } from './ubi/ubiOrganization';

export interface User {
    address: string;
    avatarMediaId: number | null;
    username: string | null;
    language: string;
    currency: string;
    pushNotificationToken: string | null;
    gender: string;
    year: number | null;
    children: number | null;
    lastLogin: Date;

    // timestamps
    createdAt: Date;
    updatedAt: Date;

    avatar?: AppMediaContent;
    // throughTrust?: AppUserTrust[];
    beneficiary?: BeneficiaryAttributes[];
    manager?: ManagerAttributes[];
}

export interface UbiCommunitySuspect {
    id: number;
    communityId: number;
    suspect: number;
    createdAt: boolean;
}
export interface UbiCommunity {
    id: number; // Note that the `null assertion` `!` is required in strict mode.
    publicId: string;
    requestByAddress: string;
    contractAddress: string | null;
    name: string;
    description: string;
    descriptionEn: string | null;
    language: string;
    currency: string;
    city: string;
    country: string;
    gps: {
        latitude: number;
        longitude: number;
    };
    email: string;
    visibility: 'public' | 'private';
    coverMediaId: number;
    status: 'pending' | 'valid' | 'removed'; // pending / valid / removed
    started: Date;

    // timestamps
    createdAt: Date;
    updatedAt: Date;

    cover?: AppMediaContent;
}
export interface CommunityAttributes extends UbiCommunity {
    metrics?: UbiCommunityDailyMetrics;
    contract?: UbiCommunityContract;
    state?: UbiCommunityState;
    suspect?: UbiCommunitySuspect;
    organization?: UbiOrganization;
}

export interface CommunityCampaing {
    communityId: number;
    campaignUrl: string;
}
export interface AppMediaThumbnail {
    id: number;
    mediaContentId: number;
    url: string;
    width: number;
    height: number;
    pixelRatio: number;
}
export interface AppMediaThumbnailCreation {
    mediaContentId: number;
    url: string;
    width: number;
    height: number;
    pixelRatio: number;
}
export interface AppMediaContent {
    id: number;
    url: string;
    width: number;
    height: number;

    thumbnails?: AppMediaThumbnail[];
}
export interface AppMediaContentCreation {
    url: string;
    width: number;
    height: number;
}

export interface CommunityStateAttributes {
    claimed: string;
    claims: number;
    beneficiaries: number;
    suspect: any[];
    removedBeneficiaries: number;
    managers: number;
    raised: string;
    backers: number;
}

export interface CommunityContractAttributes {
    claimAmount: string;
    maxClaim: string;
    baseInterval: number;
    incrementInterval: number;
}

export interface CommunityDailyMetricsAttributes {
    ssiDayAlone: number;
    ssi: number;
    ubiRate: number;
    estimatedDuration: number;
    date: Date;
}

export interface BeneficiaryAttributes {
    address: string;
    communityId: string;
    active: boolean;
    tx: string;
    txAt: Date;
    claims: number;
    claimed: string;
    blocked: boolean;
    lastClaimAt: Date | null;
    penultimateClaimAt: Date | null;
}

export interface ManagerAttributes {
    id: number;
    address: string;
    communityId: string;
    active: boolean;

    // timestamps
    createdAt: Date;
    updatedAt: Date;

    user?: User;
    community?: CommunityAttributes;
}

export interface UserAttributes {
    address: string;
    username: string | null;
    language: string;
    currency: string;
    avatar: string | null;
    // avatarMediaId: number | null;
    // pushNotificationToken: string | null;
    gender: string | null;
    year: number | null;
    children: number | null;
    blocked: boolean;
    suspect: boolean;
}

export interface IManagerAttributes {
    user: string;
    communityId: string;
}

export interface UbiRequestChangeParams {
    id: number;
    communityId: string;
    claimAmount: string;
    maxClaim: string;
    baseInterval: number;
    incrementInterval: number;
}
