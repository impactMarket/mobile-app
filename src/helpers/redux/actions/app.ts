import { ContractKit } from '@celo/contractkit';
import {
    appAction,
    SET_APP_FROM_WELCOME_SCREEN,
    SET_APP_SUSPECT_WRONG_DATETIME,
    SET_CELO_KIT,
    SET_EXCHANGE_RATES,
} from 'helpers/constants';
import { AppActionTypes } from 'helpers/types/redux';
import { Subscription } from '@unimodules/core';

export function setCeloKit(kit: ContractKit): AppActionTypes {
    return {
        type: SET_CELO_KIT,
        payload: kit,
    };
}

export function setAppSuspectWrongDateTime(
    suspect: boolean,
    timeDiff: number
): AppActionTypes {
    return {
        type: SET_APP_SUSPECT_WRONG_DATETIME,
        payload: {
            suspect,
            timeDiff,
        },
    };
}

export function SetAppFromWelcomeScreen(nextScreen: string): AppActionTypes {
    return {
        type: SET_APP_FROM_WELCOME_SCREEN,
        payload: nextScreen,
    };
}

export function setAppExchangeRatesAction(exchangeRates: any): AppActionTypes {
    return {
        type: SET_EXCHANGE_RATES,
        payload: exchangeRates,
    };
}

export function setPushNotificationListeners(notificationsListeners: {
    notificationReceivedListener: Subscription;
    notificationResponseReceivedListener: Subscription;
}): AppActionTypes {
    return {
        type: appAction.SET_PUSH_NOTIFICATION_LISTENERS,
        payload: notificationsListeners,
    };
}
