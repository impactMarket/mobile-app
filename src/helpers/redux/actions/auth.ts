import {
    SET_USER_AUTH_REQUEST,
    SET_USER_AUTH_SUCCESS,
    SET_USER_AUTH_FAILURE,
    SET_PUSH_NOTIFICATION_TOKEN,
    SET_AUTH_TOKEN,
    SET_USER_AUTH_RESET,
} from 'helpers/constants';
import { IUserAuth } from 'helpers/types/endpoints';
import { AuthActionTypes } from 'helpers/types/redux';
import { markActionsOffline } from 'redux-offline-queue';
import { AuthParams } from 'services/api/routes/user';

markActionsOffline(SET_USER_AUTH_REQUEST, [addUserAuthToStateRequest]);

export function addUserAuthToStateRequest(
    authParams: AuthParams
): AuthActionTypes {
    return {
        type: SET_USER_AUTH_REQUEST,
        payload: authParams,
    };
}

export function addUserAuthToStateSuccess(user: IUserAuth): AuthActionTypes {
    return {
        type: SET_USER_AUTH_SUCCESS,
        payload: user,
    };
}

export function addUserAuthToStateFailure(err: string): AuthActionTypes {
    return {
        type: SET_USER_AUTH_FAILURE,
        payload: err,
    };
}

export function userAuthToStateReset(): AuthActionTypes {
    return {
        type: SET_USER_AUTH_RESET,
    };
}

export function setPushNotificationsToken(token: string): AuthActionTypes {
    return {
        type: SET_PUSH_NOTIFICATION_TOKEN,
        payload: token,
    };
}

export function setAuthToken(token: string): AuthActionTypes {
    return {
        type: SET_AUTH_TOKEN,
        payload: token,
    };
}
