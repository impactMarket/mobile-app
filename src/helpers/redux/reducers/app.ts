import {
    appAction,
    SET_APP_FROM_WELCOME_SCREEN,
    SET_APP_SUSPECT_WRONG_DATETIME,
    SET_CELO_KIT,
    SET_EXCHANGE_RATES,
} from 'helpers/constants';
import { AppActionTypes } from 'helpers/types/redux';
import { IAppState } from 'helpers/types/state';

const INITIAL_STATE_APP: IAppState = {
    kit: undefined as any,
    exchangeRates: undefined as any, // save exhangeRates on load
    suspectWrongDateTime: false,
    timeDiff: 0,
    fromWelcomeScreen: '',
};

export const appReducer = (
    state = INITIAL_STATE_APP,
    action: AppActionTypes
) => {
    switch (action.type) {
        case SET_CELO_KIT:
            return { ...state, kit: action.payload };
        case SET_EXCHANGE_RATES:
            return { ...state, exchangeRates: action.payload };
        case SET_APP_SUSPECT_WRONG_DATETIME:
            return {
                ...state,
                suspectWrongDateTime: action.payload.suspect,
                timeDiff: action.payload.timeDiff,
            };
        case SET_APP_FROM_WELCOME_SCREEN:
            return {
                ...state,
                fromWelcomeScreen: action.payload,
            };
        case appAction.SET_PUSH_NOTIFICATION_LISTENERS:
            return {
                ...state,
                notificationsListeners: action.payload,
            };
        default:
            return state;
    }
};
