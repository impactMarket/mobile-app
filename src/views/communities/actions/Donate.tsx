import i18n from 'assets/i18n';
import BigNumber from 'bignumber.js';
import {
    iptcColors,
    updateCommunityInfo,
    formatInputAmountToTransfer,
    getCurrencySymbol,
} from 'helpers/index';
import { ICommunityInfo, IRootState } from 'helpers/types';
import React, { Component } from 'react';
import { StyleSheet, Clipboard, Alert, View, TextInput } from 'react-native';
import {
    // Button,
    Paragraph,
    Portal,
    Snackbar,
    Text,
    Modal,
    Headline,
    IconButton,
    // Card,
} from 'react-native-paper';
import { ConnectedProps, connect } from 'react-redux';
import { analytics } from 'services/analytics';
import { celoWalletRequest } from 'services/celoWallet';
import { writeLog } from 'services/logger/write';
import config from '../../../../config';
import * as Device from 'expo-device';
import Button from 'components/Button';
import Card from 'components/Card';

interface IExploreScreenProps {
    community: ICommunityInfo;
}
const mapStateToProps = (state: IRootState) => {
    const { user, app } = state;
    return { user, app };
};

const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;
type Props = PropsFromRedux & IExploreScreenProps;

interface IDonateState {
    openModalDonate: boolean;
    donating: boolean;
    amountDonate: string;
    showCopiedToClipboard: boolean;
    modalConfirmSend: boolean;
    rates: any;
}
BigNumber.config({ EXPONENTIAL_AT: [-7, 30] });
class Donate extends Component<Props, IDonateState> {
    constructor(props: any) {
        super(props);
        this.state = {
            openModalDonate: false,
            donating: false,
            amountDonate: '',
            showCopiedToClipboard: false,
            modalConfirmSend: false,
            rates: this.props.app.exchangeRates,
        };
    }

    handleConfirmDonateWithCeloWallet = () => {
        const { amountDonate } = this.state;
        const amountInDollars =
            parseFloat(formatInputAmountToTransfer(amountDonate)) /
            this.props.user.user.exchangeRate;
        if (
            amountInDollars >
            new BigNumber(this.props.user.celoInfo.balance)
                .dividedBy(10 ** config.cUSDDecimals)
                .toNumber()
        ) {
            Alert.alert(
                i18n.t('donate'),
                i18n.t('donationBiggerThanBalance'),
                [{ text: i18n.t('close') }],
                { cancelable: true }
            );
            return;
        }
        const { community, user } = this.props;
        Alert.alert(
            i18n.t('donate'),
            i18n.t('donateConfirmMessage', {
                symbol: getCurrencySymbol(user.user.currency),
                amount: amountDonate.replace(/\B(?=(\d{3})+(?!\d))/g, ','),
                amountInDollars: amountInDollars.toFixed(2),
                to: community.name,
            }),
            [
                {
                    text: i18n.t('donate'),
                    onPress: () => this.donateWithCeloWallet(),
                },
                {
                    text: i18n.t('cancel'),
                },
            ],
            { cancelable: true }
        );
    };

    donateWithCeloWallet = async () => {
        this.setState({ donating: true });
        const stableToken = await this.props.app.kit.contracts.getStableToken();
        const cUSDDecimals = await stableToken.decimals();
        const amountInDollars =
            parseFloat(formatInputAmountToTransfer(this.state.amountDonate)) /
            this.props.user.user.exchangeRate;
        const txObject = stableToken.transfer(
            this.props.community.contractAddress,
            new BigNumber(amountInDollars)
                .multipliedBy(new BigNumber(10).pow(cUSDDecimals))
                .toString()
        ).txo;
        celoWalletRequest(
            this.props.user.celoInfo.address,
            stableToken.address,
            txObject,
            'donatetocommunity',
            this.props.app.kit
        )
            .then((tx) => {
                console.log(tx);
                // TODO: wait for tx confirmation and request UI update
                // update donated values
                setTimeout(
                    () =>
                        updateCommunityInfo(
                            this.props.user.celoInfo.address,
                            this.props
                        ),
                    10000
                );

                Alert.alert(
                    i18n.t('success'),
                    i18n.t('youHaveDonated'),
                    [{ text: 'OK' }],
                    { cancelable: false }
                );
                analytics('donate', { device: Device.brand, success: true });
            })
            .catch((e) => {
                writeLog({ action: 'donate', details: e.message });
                analytics('donate', { device: Device.brand, success: false });
                Alert.alert(
                    i18n.t('failure'),
                    i18n.t('errorDonating'),
                    [{ text: 'OK' }],
                    { cancelable: false }
                );
            })
            .finally(() => {
                this.setState({
                    openModalDonate: false,
                    donating: false,
                    amountDonate: '',
                });
            });
    };

    handleCopyAddressToClipboard = () => {
        Clipboard.setString(this.props.community.contractAddress);
        this.setState({ showCopiedToClipboard: true });
        setTimeout(() => this.setState({ showCopiedToClipboard: false }), 5000);
        this.setState({ openModalDonate: false });
    };

    render() {
        const {
            openModalDonate,
            donating,
            amountDonate,
            showCopiedToClipboard,
            rates,
        } = this.state;
        const { community, user } = this.props;

        const amountInDollars =
            parseFloat(formatInputAmountToTransfer(amountDonate)) /
            this.props.user.user.exchangeRate;
        const amountInCommunityCurrency =
            amountInDollars * rates[community.currency].rate;

        let donatingModalString = '';
        const backForDays =
            amountInDollars /
            new BigNumber(community.vars._claimAmount)
                .dividedBy(10 ** config.cUSDDecimals)
                .toNumber() /
            community.beneficiaries.added.length;
        if (backForDays < 1) {
            donatingModalString = i18n.t('yourDonationWillBack', {
                backNBeneficiaries: Math.max(
                    1,
                    Math.floor(
                        amountInDollars /
                            new BigNumber(community.vars._claimAmount)
                                .dividedBy(10 ** config.cUSDDecimals)
                                .toNumber()
                    )
                ),
            });
        } else {
            donatingModalString = i18n.t('yourDonationWillBackFor', {
                backNBeneficiaries: community.beneficiaries.added.length,
                backForDays: Math.floor(backForDays),
            });
        }

        const donateWithValoraButton =
            user.celoInfo.address.length > 0 ? (
                <Button
                    modeType="default"
                    bold={true}
                    loading={donating}
                    disabled={donating}
                    onPress={this.handleConfirmDonateWithCeloWallet}
                >
                    {i18n.t('donateWithValora')}
                </Button>
            ) : (
                <Button
                    icon="alert"
                    modeType="default"
                    bold={true}
                    style={{
                        backgroundColor: '#f0ad4e',
                    }}
                    onPress={() => {
                        Alert.alert(
                            i18n.t('failure'),
                            i18n.t('youAreNotConnected'),
                            [{ text: i18n.t('close') }],
                            { cancelable: false }
                        );
                    }}
                >
                    {i18n.t('donateWithValora')}
                </Button>
            );

        return (
            <>
                <Button
                    modeType="green"
                    bold={true}
                    style={styles.donate}
                    labelStyle={{
                        fontSize: 20,
                        lineHeight: 23,
                        color: 'white',
                        fontWeight: 'bold',
                        // backgroundColor: 'red',
                    }}
                    onPress={() => this.setState({ openModalDonate: true })}
                >
                    {i18n.t('donate')}
                </Button>
                <Snackbar
                    visible={showCopiedToClipboard}
                    onDismiss={() => {}}
                    action={{
                        label: '',
                        onPress: () => {
                            // Do something
                        },
                    }}
                >
                    {i18n.t('addressCopiedClipboard')}
                </Snackbar>
                <Portal>
                    <Modal
                        visible={openModalDonate}
                        onDismiss={() =>
                            this.setState({ openModalDonate: false })
                        }
                    >
                        <Card style={{ marginHorizontal: 20 }}>
                            <Card.Content>
                                <View style={{ height: 40 }}>
                                    <View
                                        style={{
                                            flex: 1,
                                            flexDirection: 'row',
                                            justifyContent: 'space-between',
                                        }}
                                    >
                                        <Headline style={{ marginVertical: 3 }}>
                                            {i18n.t('donateSymbol', {
                                                symbol: user.user.currency,
                                            })}
                                        </Headline>
                                        <IconButton
                                            icon="close"
                                            size={20}
                                            style={{ top: -5 }}
                                            onPress={() =>
                                                this.setState({
                                                    openModalDonate: false,
                                                })
                                            }
                                        />
                                    </View>
                                </View>
                                <View
                                    style={{
                                        backgroundColor: '#F6F6F7',
                                        // opacity: 0.27,
                                        alignItems: 'center',
                                        borderRadius: 5,
                                    }}
                                >
                                    <TextInput
                                        keyboardType="numeric"
                                        placeholder={i18n.t('amountSymbol', {
                                            symbol: getCurrencySymbol(
                                                user.user.currency
                                            ),
                                        })}
                                        style={{
                                            fontFamily: 'Gelion-Regular',
                                            fontStyle: 'normal',
                                            fontWeight: 'normal',
                                            fontSize: 40,
                                            lineHeight: 50,
                                            textAlign: 'center',
                                        }}
                                        value={amountDonate}
                                        onChangeText={(text) =>
                                            this.setState({
                                                amountDonate: text,
                                            })
                                        }
                                    />
                                    {amountDonate.length > 0 && (
                                        <Paragraph
                                            style={{ marginVertical: 10 }}
                                        >
                                            ~
                                            {`${getCurrencySymbol(
                                                community.currency
                                            )}${
                                                (Math.floor(
                                                    amountInCommunityCurrency *
                                                        100
                                                ) / 100).toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
                                            } ${community.currency}`}
                                        </Paragraph>
                                    )}
                                </View>
                                {amountDonate.length > 0 && (
                                    <Paragraph
                                        style={{
                                            marginBottom: 20,
                                            textAlign: 'center',
                                        }}
                                    >
                                        {donatingModalString}
                                    </Paragraph>
                                )}
                                <Button
                                    modeType="gray"
                                    bold={true}
                                    style={{
                                        // backgroundColor: '#F2F3F5',
                                        marginVertical: 10,
                                    }}
                                    onPress={this.handleCopyAddressToClipboard}
                                >
                                    {i18n.t('copyContractAddress')}
                                </Button>
                                {donateWithValoraButton}
                            </Card.Content>
                        </Card>
                    </Modal>
                </Portal>
            </>
        );
    }
}

const styles = StyleSheet.create({
    donate: {
        borderRadius: 0,
        height: 69,
    },
});

export default connector(Donate);
