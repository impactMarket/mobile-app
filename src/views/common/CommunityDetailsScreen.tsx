import { Entypo } from '@expo/vector-icons';
import { useNavigation } from '@react-navigation/native';
import i18n from 'assets/i18n';
import CommuntyStatus from 'components/CommuntyStatus';
import Header from 'components/Header';
import { LinearGradient } from 'expo-linear-gradient';
import * as WebBrowser from 'expo-web-browser';
import { humanifyNumber, getCurrencySymbol, iptcColors } from 'helpers/index';
import {
    ICommunityInfo,
    IStoreCombinedActionsTypes,
    IStoreCombinedState,
} from 'helpers/types';
import React, { useState } from 'react';
import { View, StyleSheet, RefreshControl, Image } from 'react-native';
import { ScrollView } from 'react-native-gesture-handler';
import { Paragraph, Divider, Headline, Text } from 'react-native-paper';

import config from '../../../config';
import Donate from '../communities/Donate';
import Api from 'services/api';
import { useStore } from 'react-redux';
import Button from 'components/Button';
import Card from 'components/Card';

import { LineChart } from 'react-native-svg-charts';
import * as shape from 'd3-shape';
import BaseCommunity from 'components/BaseCommunity';

interface ICommunityDetailsScreen {
    route: {
        params: {
            community: ICommunityInfo;
        };
    };
}
export default function CommunityDetailsScreen(props: ICommunityDetailsScreen) {
    const store = useStore<IStoreCombinedState, IStoreCombinedActionsTypes>();
    const navigation = useNavigation();
    const rates = store.getState().app.exchangeRates;

    const [refreshing, setRefreshing] = useState(false);
    const [seeFullDescription, setSeeFullDescription] = useState(false);
    const [community, setCommunity] = useState<ICommunityInfo>(
        props.route.params.community
    );

    const onRefresh = () => {
        Api.getCommunityByContractAddress(community.contractAddress).then((c) =>
            setCommunity(c!)
        );
        setRefreshing(false);
    };

    const renderSSI = () => {
        if (community.ssi.values.length > 1) {
            return (
                <>
                    <Divider
                        style={{
                            marginLeft: -16,
                            marginRight: -16,
                            marginTop: 16,
                        }}
                    />
                    <View style={styles.chartView}>
                        <LineChart
                            style={{ flex: 2, height: 100, width: 200 }}
                            data={community.ssi.values}
                            contentInset={{ top: 20, bottom: 20 }}
                            curve={shape.curveMonotoneX}
                            svg={{
                                strokeWidth: 4,
                                stroke: 'rgba(45,206,137,1)',
                            }}
                        ></LineChart>
                        <View
                            style={{
                                ...styles.ssiView,
                                // backgroundColor: 'blue',
                            }}
                        >
                            <View
                                style={{
                                    // backgroundColor: 'red',
                                    // flex: 1,
                                    flexDirection: 'row',
                                    justifyContent: 'flex-end',
                                }}
                            >
                                <Headline style={styles.ssiHeadline}>
                                    {
                                        community.ssi.values[
                                            community.ssi.values.length - 1
                                        ]
                                    }
                                </Headline>
                                <Text
                                    style={{
                                        textAlignVertical: 'center',
                                        fontSize: 18,
                                        lineHeight: 18,
                                        padding: 3,
                                    }}
                                >
                                    %
                                </Text>
                            </View>
                            <Paragraph
                                style={{
                                    textAlign: 'right',
                                    fontSize: 14,
                                    lineHeight: 17,
                                }}
                            >
                                {i18n.t('ssi')}
                            </Paragraph>
                        </View>
                    </View>
                    <Paragraph style={styles.ssiExplained}>
                        {i18n.t('ssiDescription')}
                    </Paragraph>
                </>
            );
        }
    };

    let description;
    const cDescription =
        store.getState().user.user.language === community.language
            ? community.description
            : community.descriptionEn === null
            ? community.description
            : community.descriptionEn;
    if (seeFullDescription || community.description.indexOf('\n') === -1) {
        description = cDescription;
    } else {
        description = cDescription.slice(0, cDescription.indexOf('\n'));
    }
    const amountInDollars = parseFloat(community.vars._claimAmount);
    const amountInCommunityCurrency =
        amountInDollars * rates[community.currency].rate;

    return (
        <>
            <Header title="" hasBack hasHelp navigation={navigation} />
            <ScrollView
                refreshControl={
                    <RefreshControl
                        refreshing={refreshing}
                        onRefresh={onRefresh}
                    />
                }
            >
                <BaseCommunity community={community}>
                    <View style={styles.container}>
                        <Card elevation={0}>
                            <Card.Content>
                                <Paragraph
                                    style={{
                                        fontSize: 17,
                                        lineHeight: 22,
                                    }}
                                >
                                    {description}
                                </Paragraph>
                                {community.description.indexOf('\n') !== -1 && (
                                    <Button
                                        modeType="gray"
                                        onPress={() =>
                                            setSeeFullDescription(
                                                !seeFullDescription
                                            )
                                        }
                                    >
                                        {seeFullDescription
                                            ? i18n.t('seeLess')
                                            : i18n.t('seeMore')}
                                    </Button>
                                )}
                            </Card.Content>
                        </Card>
                        <Card elevation={0} style={{ marginTop: 16 }}>
                            <Card.Content>
                                <Paragraph
                                    style={{
                                        fontSize: 17,
                                        lineHeight: 21,
                                    }}
                                >
                                    {i18n.t('eachBeneficiaryCanClaimXUpToY', {
                                        communityCurrency: getCurrencySymbol(
                                            community.currency
                                        ),
                                        claimXCCurrency: humanifyNumber(
                                            amountInCommunityCurrency.toString()
                                        ),
                                        claimX: humanifyNumber(
                                            community.vars._claimAmount
                                        ),
                                        upToY: humanifyNumber(
                                            community.vars._maxClaim
                                        ),
                                        minIncrement:
                                            parseInt(
                                                community.vars
                                                    ._incrementInterval
                                            ) / 60,
                                    })}
                                </Paragraph>
                                {renderSSI()}
                            </Card.Content>
                        </Card>
                        <CommuntyStatus community={community}>
                            <Button
                                modeType="gray"
                                bold={true}
                                style={{ marginTop: '5%' }}
                                onPress={() =>
                                    WebBrowser.openBrowserAsync(
                                        config.blockExplorer +
                                            community.contractAddress +
                                            '/token_transfers'
                                    )
                                }
                            >
                                {i18n.t('exploreCommunityContract')}
                            </Button>
                        </CommuntyStatus>
                    </View>
                </BaseCommunity>
            </ScrollView>
            <Donate community={community} />
        </>
    );
}

const styles = StyleSheet.create({
    container: {
        marginHorizontal: 16,
        marginBottom: 30,
    },
    imageBackground: {
        width: '100%',
        height: 152,
        justifyContent: 'center',
        alignContent: 'center',
        alignItems: 'center',
    },
    linearGradient: {
        position: 'absolute',
        left: 0,
        right: 0,
        bottom: 0,
        height: 25,
    },
    darkerBackground: {
        position: 'absolute',
        left: 0,
        right: 0,
        bottom: 0,
        height: 152,
    },
    communityName: {
        zIndex: 5,
        marginTop: 22,
        marginBottom: 2,
        fontSize: 30,
        lineHeight: 36,
        fontFamily: 'Gelion-Bold',
        color: 'white',
        textAlign: 'center',
    },
    communityLocation: {
        zIndex: 5,
        marginBottom: 32.41,
        fontSize: 15,
        lineHeight: 15,
        letterSpacing: 0.25,
        color: 'white',
        textAlign: 'center',
    },
    ssiExplained: {
        fontSize: 15,
        lineHeight: 18,
        letterSpacing: 0.25,
        color: iptcColors.textGray,
    },
    ssiHeadline: {
        fontFamily: 'Gelion-Bold',
        fontSize: 36,
        lineHeight: 36,
        textAlign: 'right',
    },
    ssiView: {
        flex: 2,
        flexDirection: 'column',
        justifyContent: 'center',
        marginRight: 0,
    },
    chartView: {
        flex: 3,
        flexDirection: 'row',
        margin: 0,
    },
});
