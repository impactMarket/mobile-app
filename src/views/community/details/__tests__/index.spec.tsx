import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
import { act, fireEvent, render } from '@testing-library/react-native';
import { setAppExchangeRatesAction } from 'helpers/redux/actions/app';
import combinedReducer from 'helpers/redux/reducers';
import { CommunityAttributes } from 'helpers/types/models';
import React from 'react';
import { Host } from 'react-native-portalize';
import * as reactRedux from 'react-redux';
import { createStore } from 'redux';
import Api from 'services/api';
import CommunityExtendedDetailsScreen from 'views/community/extendedDetails';

import CommunityDetailsScreen from '../';

// TODO: why is useDispatch behaving different from create community?
jest.mock('react-redux', () => ({
    ...(jest.requireActual('react-redux') as any),
    useSelector: jest.fn(),
    // useDispatch: jest.fn(),
    batch: jest.fn(),
}));

jest.mock('components/community/Description');

const community: CommunityAttributes = {
    id: 1,
    publicId: '3f49b131-e097-4155-8f28-46c95590d42f',
    requestByAddress: '0xb10199414d158a264e25a5ec06b463c0cd8457bb',
    contractAddress: '0xd48466ffc8b6190d33fb6b27035a032658d392ee',
    name: 'Example Community',
    description:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse quis interdum justo, nec sollicitudin lectus. Quisque malesuada metus lacinia facilisis pretium. Maecenas semper lacus id ante porttitor, non ultricies nunc suscipit. Vestibulum hendrerit eros vel mollis blandit. Nam tincidunt libero nunc, sed gravida tortor vulputate ut. Donec ullamcorper dui erat, vitae consectetur neque auctor sit amet. Aliquam a lacus congue, scelerisque orci vitae, accumsan tellus. Nunc ultrices convallis euismod.',
    city: 'Lauro de Freitas',
    country: 'BR',
    gps: { latitude: 1, longitude: 2 },
    email: 'eu@amigos.txt',
    visibility: 'public',
    status: 'valid',
    createdAt: new Date(),
    updatedAt: new Date(),
    currency: 'BRL',
    descriptionEn: null,
    language: 'pt',
    started: new Date(),
    coverMediaId: 260,
    contract: {
        claimAmount: '750000000000000000',
        maxClaim: '400000000000000000000',
        baseInterval: 86400,
        incrementInterval: 300,
        createdAt: new Date(),
        updatedAt: new Date(),
        communityId: 1,
    },
    state: {
        claimed: '61942500000000000000000',
        claims: 55405,
        beneficiaries: 656,
        raised: '62151145240504400207738',
        backers: 37,
        createdAt: new Date(),
        updatedAt: new Date(),
        removedBeneficiaries: 668,
        managers: 4,
        communityId: 1,
    },
    cover: {
        id: 2,
        url: 'https://dzrx8kf1cwjv9.cloudfront.net/cover/1620119094527.jpeg',
        width: 800,
        height: 800,
        thumbnails: [
            {
                id: 1,
                mediaContentId: 2,
                url: 'https://dzrx8kf1cwjv9.cloudfront.net/cover/330x330/1620119095297.jpeg',
                width: 330,
                height: 330,
                pixelRatio: 1,
            },
        ],
    },
};

/**
 * NOTE: we are testing the component individually, but need the header
 * "submit button", so the entire navigator can be faked.
 */
function WrappedCommunityDetailsScreen() {
    const Stack = createStackNavigator();
    const store = createStore(combinedReducer);
    store.dispatch(setAppExchangeRatesAction({ USD: 1, BRL: 0.3 }));
    return (
        <Host>
            <reactRedux.Provider store={store}>
                <NavigationContainer>
                    <Stack.Navigator>
                        <Stack.Screen
                            name="Home"
                            component={CommunityDetailsScreen}
                            initialParams={{
                                route: {
                                    params: {
                                        communityId: 1,
                                    },
                                },
                            }}
                        />
                        <Stack.Screen
                            name="CommunityExtendedDetailsScreen"
                            component={CommunityExtendedDetailsScreen}
                            initialParams={{
                                route: {
                                    params: {
                                        communityId: 1,
                                    },
                                },
                            }}
                        />
                    </Stack.Navigator>
                </NavigationContainer>
            </reactRedux.Provider>
        </Host>
    );
}

describe('details [snapshot]', () => {
    const useSelectorMock = reactRedux.useSelector as jest.Mock<any, any>;

    beforeAll(() => {
        useSelectorMock.mockImplementation((callback) => {
            return callback({
                app: {
                    exchangeRates: { USD: 1 },
                },
                user: {
                    metadata: {
                        currency: 'USD',
                        avatar: 'something.jpg',
                    },
                    wallet: {
                        address: '0xd7632B7588DF8532C0aBA55586167C2a315Fd768',
                    },
                },
                communities: {
                    community,
                },
            });
        });
    });

    it('renders correctly', async () => {
        const tree = render(<WrappedCommunityDetailsScreen />);
        await act(async () => {});
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('open "donate with esolidar when URL is present"', async () => {
        const communityFundraisingUrl = jest.spyOn(
            Api.community,
            'getCampaign'
        );

        communityFundraisingUrl.mockImplementation(() =>
            Promise.resolve({
                communityId: 1,
                campaignUrl: 'https://community.esolidar.com/pt/fake',
            })
        );

        const { getByTestId, queryByTestId } = render(
            <WrappedCommunityDetailsScreen />
        );
        await act(async () => {});

        fireEvent.press(getByTestId('donateWithESolidar'));
        expect(queryByTestId('webViewESolidar')).not.toBeNull();
    });

    it('dont show "donate with esolidar button when URL isnt present"', async () => {
        const communityFundraisingUrl = jest.spyOn(
            Api.community,
            'getCampaign'
        );

        communityFundraisingUrl.mockImplementation(() =>
            Promise.resolve({
                communityId: 1,
                campaignUrl: null,
            })
        );

        const { queryByTestId } = render(<WrappedCommunityDetailsScreen />);
        await act(async () => {});

        expect(queryByTestId('donateWithESolidar')).toBeNull();
    });

    it('open "donate with celo dollar"', async () => {
        const { getByTestId, queryByTestId } = render(
            <WrappedCommunityDetailsScreen />
        );
        await act(async () => {});

        fireEvent.press(getByTestId('donateWithCelo'));
        expect(queryByTestId('modalDonateWithCelo')).not.toBeNull();
    });

    // TODO: open and close "donate with esolidar"
    // TODO: open and close "donate with celo dollar"

    it('press "see more"', async () => {
        const { getByTestId, queryByTestId } = render(
            <WrappedCommunityDetailsScreen />
        );
        await act(async () => {});

        fireEvent.press(getByTestId('communitySeeMore'));

        expect(queryByTestId('communityExtendedDetails')).not.toBeNull();
    });

    // TODO: test donating with success
    // TODO: test donating with higher amount that wallet balance
    // TODO: test cancel donate process and start over again
});
