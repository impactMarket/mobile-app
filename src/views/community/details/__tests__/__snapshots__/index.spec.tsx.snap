// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Community Details screen test suite should render CommunityDetails screen correctly 1`] = `ShallowWrapper {}`;
